helix-cli/src/metrics_sender.rs

```diff
@@ -0,0 +1,413 @@
+            install_event_sent: false,
+        }
+    }
+}
+
+pub struct MetricsSender {
+    tx: Sender<MetricsMessage>,
+    handle: JoinHandle<()>,
+}
+
+#[derive(Debug)]
+enum MetricsMessage {
+    Event(RawEvent<EventData>),
+    Shutdown,
+}
+
+impl MetricsSender {
+    pub fn new() -> Result<Self> {
+        let (tx, rx) = unbounded();
+        let handle = tokio::spawn(async move {
+            if let Err(e) = metrics_task(rx).await {
+                eprintln!("Metrics task error: {e}");
+            }
+        });
+
+        Ok(Self { tx, handle })
+    }
+
+    pub fn send_event(&self, event: RawEvent<EventData>) {
+        let _ = self.tx.send(MetricsMessage::Event(event));
+    }
+
+    pub async fn shutdown(self) -> Result<()> {
+        let _ = self.tx.send(MetricsMessage::Shutdown);
+        self.handle
+            .await
+            .map_err(|e| eyre!("Metrics task join error: {e}"))?;
+        Ok(())
+    }
+}
+
+async fn metrics_task(rx: Receiver<MetricsMessage>) -> Result<()> {
+    let mut log_writer = None;
+
+    let config = load_metrics_config().unwrap_or_default();
+
+    if config.level != MetricsLevel::Off {
+        let _ = upload_previous_logs().await;
+
+        let metrics_dir = get_metrics_dir()?;
+        let today = Local::now().format("%Y-%m-%d").to_string();
+        let log_file_path = metrics_dir.join(format!("{today}.json"));
+
+        log_writer = create_log_writer(&log_file_path).ok();
+    }
+
+    while let Ok(message) = rx.recv_async().await {
+        match message {
+            MetricsMessage::Event(event) => {
+                if let Some(ref mut writer) = log_writer
+                    && let Err(e) = write_event_to_log(writer, &event)
+                {
+                    eprintln!("Failed to write metrics event: {e}");
+                }
+            }
+            MetricsMessage::Shutdown => {
+                break;
+            }
+        }
+    }
+
+    if let Some(mut writer) = log_writer {
+        let _ = writer.flush();
+    }
+
+    Ok(())
+}
+
+pub(crate) fn load_metrics_config() -> Result<MetricsConfig> {
+    let config_path = get_metrics_config_path()?;
+
+    if !config_path.exists() {
+        return Ok(MetricsConfig::default());
+    }
+
+    let content = fs::read_to_string(&config_path)?;
+    let config = toml::from_str(&content)?;
+    Ok(config)
+}
+
+pub(crate) fn save_metrics_config(config: &MetricsConfig) -> Result<()> {
+    let config_path = get_metrics_config_path()?;
+    let content = toml::to_string_pretty(config)?;
+    fs::write(&config_path, content)?;
+    Ok(())
+}
+
+pub(crate) fn get_metrics_config_path() -> Result<PathBuf> {
+    let home = home_dir().ok_or_else(|| eyre!("Cannot find home directory"))?;
+    let helix_dir = home.join(".helix");
+    fs::create_dir_all(&helix_dir)?;
+    Ok(helix_dir.join("metrics.toml"))
+}
+
+fn get_metrics_dir() -> Result<PathBuf> {
+    let home = home_dir().ok_or_else(|| eyre!("Cannot find home directory"))?;
```

Comment:
Everywhere there is an `.ok_or_else(|| eyre!`, if the eyre macro doesn't take any format arguments and it can be a static string, you can replace it with `.ok_or_eyre(`